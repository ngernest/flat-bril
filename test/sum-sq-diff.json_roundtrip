{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v0"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v2"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "v5"
          ],
          "funcs": [],
          "labels": [
            "for.body.1",
            "for.end.1"
          ],
          "op": "br"
        },
        {
          "label": "for.body.1"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v8"
          ],
          "dest": "sq",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "sq"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v13",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v14"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "sumOfSquares",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v0"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v2"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "v5"
          ],
          "funcs": [],
          "labels": [
            "for.body.1",
            "for.end.1"
          ],
          "op": "br"
        },
        {
          "label": "for.body.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v6",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v8"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v10",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.1"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "res"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v14"
          ],
          "dest": "square",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "square"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "squareOfSum",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0"
          ],
          "dest": "sum",
          "funcs": [
            "sumOfSquares"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "sum"
          ],
          "dest": "sum",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "dest": "square",
          "funcs": [
            "squareOfSum"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "square"
          ],
          "dest": "square",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "square"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "sum"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v4"
          ],
          "dest": "diff",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "diff"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v5"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v6",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "diff"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        }
      ],
      "name": "main"
    }
  ]
}
