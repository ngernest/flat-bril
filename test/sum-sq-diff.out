@sumOfSquares
	op: const		dest: v0		type: int		value: 0	
	op: id   		dest: res		type: int		args: ["v0"]	
	op: const		dest: v2		type: int		value: 1	
	op: id   		dest: i 		type: int		args: ["v2"]	
.for.cond.1
	op: id   		dest: v3		type: int		args: ["i"]	
	op: id   		dest: v4		type: int		args: ["n"]	
	op: le   		dest: v5		type: bool		args: ["v3", "v4"]	
	op: br   		args: ["v5"]		labels: ["for.body.1", "for.end.1"]
.for.body.1
	op: id   		dest: v6		type: int		args: ["i"]	
	op: id   		dest: v7		type: int		args: ["i"]	
	op: mul  		dest: v8		type: int		args: ["v6", "v7"]	
	op: id   		dest: sq		type: int		args: ["v8"]	
	op: id   		dest: v9		type: int		args: ["res"]	
	op: id   		dest: v10		type: int		args: ["sq"]	
	op: add  		dest: v11		type: int		args: ["v9", "v10"]	
	op: id   		dest: res		type: int		args: ["v11"]	
	op: id   		dest: v12		type: int		args: ["i"]	
	op: const		dest: v13		type: int		value: 1	
	op: add  		dest: v14		type: int		args: ["v12", "v13"]	
	op: id   		dest: i 		type: int		args: ["v14"]	
	op: jmp  		labels: ["for.cond.1"]
.for.end.1
	op: id   		dest: v15		type: int		args: ["res"]	
	op: ret  		args: ["v15"]	
@squareOfSum
	op: const		dest: v0		type: int		value: 0	
	op: id   		dest: res		type: int		args: ["v0"]	
	op: const		dest: v2		type: int		value: 1	
	op: id   		dest: i 		type: int		args: ["v2"]	
.for.cond.1
	op: id   		dest: v3		type: int		args: ["i"]	
	op: id   		dest: v4		type: int		args: ["n"]	
	op: le   		dest: v5		type: bool		args: ["v3", "v4"]	
	op: br   		args: ["v5"]		labels: ["for.body.1", "for.end.1"]
.for.body.1
	op: id   		dest: v6		type: int		args: ["res"]	
	op: id   		dest: v7		type: int		args: ["i"]	
	op: add  		dest: v8		type: int		args: ["v6", "v7"]	
	op: id   		dest: res		type: int		args: ["v8"]	
	op: id   		dest: v9		type: int		args: ["i"]	
	op: const		dest: v10		type: int		value: 1	
	op: add  		dest: v11		type: int		args: ["v9", "v10"]	
	op: id   		dest: i 		type: int		args: ["v11"]	
	op: jmp  		labels: ["for.cond.1"]
.for.end.1
	op: id   		dest: v12		type: int		args: ["res"]	
	op: id   		dest: v13		type: int		args: ["res"]	
	op: mul  		dest: v14		type: int		args: ["v12", "v13"]	
	op: id   		dest: square		type: int		args: ["v14"]	
	op: id   		dest: v15		type: int		args: ["square"]	
	op: ret  		args: ["v15"]	
@main
	op: id   		dest: v0		type: int		args: ["n"]	
	op: call 		dest: sum		type: int		args: ["v0"]		funcs: ["sumOfSquares"]
	op: id   		dest: sum		type: int		args: ["sum"]	
	op: id   		dest: v1		type: int		args: ["n"]	
	op: call 		dest: square		type: int		args: ["v1"]		funcs: ["squareOfSum"]
	op: id   		dest: square		type: int		args: ["square"]	
	op: id   		dest: v2		type: int		args: ["square"]	
	op: id   		dest: v3		type: int		args: ["sum"]	
	op: sub  		dest: v4		type: int		args: ["v2", "v3"]	
	op: id   		dest: diff		type: int		args: ["v4"]	
	op: id   		dest: v5		type: int		args: ["diff"]	
	op: print		args: ["v5"]	
	op: const		dest: v6		type: int		value: 0	
	op: id   		dest: v7		type: int		args: ["diff"]	
